pkgs/development/libraries/glibc/ldconfig-patchelf.patch

```diff
@@ -0,0 +1,67 @@
+diff --git i/elf/readelflib.c w/elf/readelflib.c
+index 5a1e2dc2df..66fef2dd6d 100644
+--- i/elf/readelflib.c
++++ w/elf/readelflib.c
+@@ -98,11 +98,6 @@ process_elf_file (const char *file_name, const char *lib, int *flag,
+ 
+       switch (segment->p_type)
+ 	{
+-	case PT_LOAD:
+-	  if (loadaddr == (ElfW(Addr)) -1)
+-	    loadaddr = segment->p_vaddr - segment->p_offset;
+-	  break;
+-
+ 	case PT_DYNAMIC:
+ 	  if (dynamic_addr)
+ 	    error (0, 0, _("more than one dynamic segment\n"));
+@@ -176,11 +171,6 @@ process_elf_file (const char *file_name, const char *lib, int *flag,
+ 	}
+ 
+     }
+-  if (loadaddr == (ElfW(Addr)) -1)
+-    {
+-      /* Very strange. */
+-      loadaddr = 0;
+-    }
+ 
+   /* Now we can read the dynamic sections.  */
+   if (dynamic_size == 0)
+@@ -190,21 +180,30 @@ process_elf_file (const char *file_name, const char *lib, int *flag,
+   check_ptr (dynamic_segment);
+ 
+   /* Find the string table.  */
+-  dynamic_strings = NULL;
+   for (dyn_entry = dynamic_segment; dyn_entry->d_tag != DT_NULL;
+        ++dyn_entry)
+     {
+       check_ptr (dyn_entry);
+       if (dyn_entry->d_tag == DT_STRTAB)
+-	{
+-	  dynamic_strings = (char *) (file_contents + dyn_entry->d_un.d_val - loadaddr);
+-	  check_ptr (dynamic_strings);
+-	  break;
+-	}
++        break;
+     }
+ 
+-  if (dynamic_strings == NULL)
+-    return 1;
++  for (i = 0, segment = elf_pheader;i < elf_header->e_phnum; i++, segment++)
++    {
++	  ElfW(Addr) vaddr = dyn_entry->d_un.d_ptr;
++	  if (segment->p_type == PT_LOAD &&
++			  vaddr >= segment->p_vaddr &&
++			  vaddr < segment->p_vaddr +segment->p_filesz)
++	  {
++		  loadaddr = segment->p_vaddr - segment->p_offset;
++		  break;
++	  }
++    }
++  if (loadaddr == (ElfW(Addr)) -1)
++	  return 1;
++
++  dynamic_strings = (char *) (file_contents + dyn_entry->d_un.d_val - loadaddr);
++  check_ptr (dynamic_strings);
```

Comment:
Is this something the glibc upstreamer could adopt?

Reply:
It's possible. I don't know if it's actually legal to put dynstr in a different segment, and I doubt a compiler would do it. Also, they have like a 10 page contribution checklist, but yes it should be tried.

Reply:
The contribution checklist looks doable as it is mostly how to format patches: https://sourceware.org/glibc/wiki/Contribution%20checklist

I can also submit a patch, but I would need a proper explanation of the change as required here: https://sourceware.org/glibc/wiki/Contribution%20checklist#Detailed_Explanation_of_the_Patch

Reply:
Alright, I've submitted a bug report: https://sourceware.org/bugzilla/show_bug.cgi?id=23964
and a patch to the mailing list, we'll see what happens.